crates/core/src/energy.rs

```diff
@@ -0,0 +1,127 @@
+use derive_more::{Add, AddAssign, From, Sub, SubAssign};
+use std::fmt;
+use std::time::Duration;
+
+use spacetimedb_lib::{Hash, Identity};
+
+use crate::messages::control_db::Database;
+
+/// [EnergyQuanta] represents an amount of energy in a canonical unit.
+/// It represents the smallest unit of energy that can be used to pay for
+/// a reducer invocation. We will likely refer to this unit as an "eV".
+///
+/// NOTE: This is represented by a signed integer, because it is possible
+/// for a user's balance to go negative. This is allowable
+/// for reasons of eventual consistency motivated by performance.
```

Comment:
> may as well just not run the reducer at all

This is true. But if ReducerBudget corresponds to an EnergyBalance, then it _could_ be signed since EnergyBalance is signed.

But the bigger question is why is EnergyQuanta signed? If EnergyQuanta is how we measure energy consumption, it will never be negative.

Although I don't know how much it will expand the diff. But I figure if you're already making the corresponding change to ReducerBudget, then you might as well do the same thing for EnergyQuanta. WDYT?

Reply:
Well, there's the comment that explains why it's signed, but yeah, I figure EnergyQuanta being unsigned would make a ton of things much simpler, and maybe we could have an EnergyBalance that's signed. I was hesitant to do that in this PR, cause I didn't want to cause too much churn, but if you also think it'd be good I can definitely do it.